### ISSUE TITLE
img-req-alt check conflates presence of alt attribute with *length* of alt text

### ISSUE THREAD

**Issue by andymantell** [htmllint/htmllint#148]
I have just updated to the most recent version of htmllint and now my tests are failing with `the `alt` property must be set for image tags`. It seems that some changes were made to this test recently which now means it checks for empty alt attributes and not just the presence of it (https://github.com/htmllint/htmllint/commit/45ddd73fdca621911a9b486066b3c3c8704dfb42)

I don't think this should be combined in a single check. I want to ensure that alt attributes are present (With `img-req-alt`) but empty alt texts should be permissible under this test because that is the correct thing to do if the image is decorational. I think that enforcing non empty alt texts should be a separate test.

Was there some other rationale for adding alt text length checks to `img-req-alt` ? Would it be worth splitting this out into a new test which checks for zero length alt texts?

**Comment by mlochbaum** [comment]
Having a non-empty alt text for each image is an accessibility requirement, but you're right that it shouldn't always be necessary. I think the correct interface is to change `img-req-alt` from a boolean to a string option. `true` or "non-empty" will enable the current behavior, and something like "required" will require an alt tag but not force it to have a value.

**Comment by andymantell** [comment]
WCAG does allow for null / zero length alt texts to indicate that a screen reader should ignore the image. See https://www.w3.org/TR/WCAG20-TECHS/H67.html

Your proposed solution sounds good though and would cater for either requirement.
Having a non-empty alt text for each image is an accessibility requirement, but you're right that it shouldn't always be necessary. I think the correct interface is to change `img-req-alt` from a boolean to a string option. `true` or "non-empty" will enable the current behavior, and something like "required" will require an alt tag but not force it to have a value.

---

**Comment by andymantell** [comment]
Sorry for the quoted email text, have edited it now...

Anyway, will see if I can find a minute tomorrow to do this and open a PR

**Comment by mlochbaum** [comment]
Fixed it for you (it's nice when changes are this easy...). I settled on "allownull" for the new option name, but I can change it if you have a better idea.

**Comment by andymantell** [comment]
Fantastic, thanks! I'll pull this in later and give it a whirl.

**Comment by AntonioRedondo** [comment]
@mlochbaum the quick fix you implemented is great. Thanks! I experienced the same issue with `img-req-alt` but your fix is perfect. However I see changes hasn't be released to NPM yet. Any idea when this will be done?

**Comment by mlochbaum** [comment]
There are some changes I want to make and am trying to finish this weekend, but I'll make sure to publish by Sunday night.
